[Modular] support standard repo

yiyixuxu opened the pull request:
fix https://github.com/huggingface/diffusers/issues/11915

this PR support using standard repo in Modular system

#### option1
```py
from diffusers import ModularPipeline
import torch

pipe = ModularPipeline.from_pretrained("stabilityai/stable-diffusion-xl-base-1.0", torch_dtype=torch.float16)
print(pipe)
```

#### option2

```py
from diffusers import StableDiffusionXLAutoBlocks
modular_pipe = StableDiffusionXLAutoBlocks().init_pipeline("stabilityai/stable-diffusion-xl-base-1.0")
modular_pipe.load_default_components(torch_dtype=torch.float16)
```

#### option3

```py
from diffusers import DiffusionPipeline, StableDiffusionXLAutoBlocks
import diffusers

modular_map= {
    'StableDiffusionXLPipeline': 'StableDiffusionXLAutoBlocks',
}

standard_pipe = DiffusionPipeline.from_pretrained("stabilityai/stable-diffusion-xl-base-1.0", torch_dtype=torch.float16)
modular_cls = modular_map.get(standard_pipe.__class__.__name__)
modular_cls = getattr(diffusers, modular_cls)
blocks = modular_cls()
modular_pipe = blocks.init_pipeline()

components_to_update = {k: v for k, v in standard_pipe.components.items() if v is not None}
modular_pipe.update_components(**components_to_update, **standard_pipe.parameters)
```

yiyixuxu commented:
cc @vladmandic let me know if this works for you

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/diffusers/pr_11944). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.

yiyixuxu commented:
@vladmandic 
#### warnings
I removed the warnings during the generation (changed to `logger.info`), but kept the ones you get during the conversion. This is because currently the expected behavior for `update_components` is to update the underlying loading spec as you update the component itself, so that we should warn users if not the case

However, we are still iterating everything, so that may change in future too

```
ModularPipeline.update_components: text_encoder has no valid _diffusers_load_id. Updating the component but skipping spec update, use ComponentSpec.load() for proper specs
```

#### image difference 

The results from modular and standard pipelines are not going to be the same: Other than the guider, we are also taking this opportunity to refactor all our pipeline methods to make them more modular and easy to customize upon - this is an ongoing effort, we are likely to change more as we integrate more pipelines into modular diffusers 

However, once modular is out of experimental feature, moving forward, we will make sure new pipelines will have same results in both systems

for now, I think if reproducibility is important for SD.NEXT users, I think we can only convert if they need to use a new guider class that's not supported in standard pipeline


#### output

I updated `PipelineState` so now you can access `images` or anything else else in intermediate states just by using that attribute; now below 3 methods to get the images outputs are all supported

```python
image = t2i_pipe(prompt=prompt, num_inference_steps=25, output="images")[0]
image.save(f"{output_name}_1.png")


image = t2i_pipe(prompt=prompt, num_inference_steps=25).images[0]
image.save(f"{output_name}_2.png")

image = t2i_pipe(prompt=prompt, num_inference_steps=25).intermediates['images'][0]
image.save(f"{output_name}_3.png")
```

vladmandic commented:
thanks @yiyixuxu 

re: warnings - all good
re: output - perfect
re: image differences
the biggest reason to start using new modular pipelines is new guider concept, so i don't mind combining the two on my side  
and reproducibility does not have to be exact using standard vs modular pipeline defaults  
just need to know how to map the params from standard to modular pipeline to achieve reproducibility
